File: 1342.number-of-steps-to-reduce-a-number-to-zero.cpp
/*
 * @lc app=leetcode id=1342 lang=cpp
 *
 * [1342] Number of Steps to Reduce a Number to Zero
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int numberOfSteps(int num) {
    int ret = 0;
    while (num) {
      if (num % 2) {
        num--;
      } else {
        num /= 2;
      }
      ret++;
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  ASSERT_EQ(s.numberOfSteps(14), 6);
  ASSERT_EQ(s.numberOfSteps(8), 4);
  return 0;
}

File: 1342.number-of-steps-to-reduce-a-number-to-zero.rs
/*
 * @lc app=leetcode id=1342 lang=rust
 *
 * [1342] Number of Steps to Reduce a Number to Zero
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn number_of_steps(num: i32) -> i32 {
        if num == 0 {
            return 0;
        }
        (num.count_ones() + 31 - num.leading_zeros()) as i32
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::number_of_steps(14), 6);
}

#[test]
fn test2() {
    assert_eq!(Solution::number_of_steps(8), 4);
}
